doc/specs/Proto extensions-spec.md

```diff
@@ -0,0 +1,378 @@
+to add json snippets to our json files, and will be used when we generate settings for our various profiles. 
+
+## Inspiration
+
+### Goal: Allow programs to have a say in the settings for their profiles
+
+Currently, Ubuntu/WSL/Powershell etc are unable to provide any specifications for how they want
+their profiles in Terminal to look - only we and the users can modify the profiles. We want to provide
+these installations with the functionality to have a say in this, allowing them to specify things like
+their icon, their font and so on. However, we want to maintain that the user has final say over all of
+these settings. 
+
+## Solution Design
+
+Currently, when we load the settings we perform the following steps (this is a simplified description,
+for the full version see the [spec for cascading default + user settings](https://github.com/microsoft/terminal/blob/master/doc/specs/%23754%20-%20Cascading%20Default%20Settings.md)):
+
+1. Generate profiles from the defaults file
+2. Generate profiles from the dynamic profile generators
+3. Layer the user settings on top of all the profiles created in steps 1 and 2
+4. Validate the settings
+
+To allow for installations to add in their snippets of json, I propose the addition of a new step
+in between 2 and 3:
+
+1. Generate profiles from the defaults file
+2. Generate profiles from the dynamic profile generators
+3. **Incorporate the additional provided json stubs** - these stubs could be:
+   1. modifications to existing profiles
+   2. additions of full profiles
+   3. additions of colour schemes
+4. Layer the user settings on top of all the profiles created in steps 1 through 3 (this includes the user-defined default settings which they wish to apply to all profiles)
+5. Validate the settings
+
+### Specifications of the json stubs
+
+As written above, the json stubs could be modifications to existing profiles, additions to full profiles, or additions of colour schemes.
+
+To allow modifications/additions of several profiles in one file and to allow the addition of several colour schemes in one json file,
+these stubs should be put into lists in the json file: one list named ```profiles``` and the other list named ```schemes```. For example:
+
+```js
+{
+    "profiles": [ modifications and additions of profiles go here ],
+    "schemes": [ additions of colour schemes go here ]
+}
+```
+
+An example of a full json file with these fields filled out is shown at the end of this section. 
+
+#### Modifications to existing profiles
+
+The main thing to note for modification of existing profiles is that this will only be used for modifying the
+default profiles (cmd/powershell) or the dynamically generated profiles. 
+
+For modifications to existing profiles, the json stub would need to indicate which profile it wishes to modify. It will
+do this by providing the corresponding guid in the `"updates"` field of the json stub.

```

Comment:
this might need to be called out in the spec better.

There's a distinct difference between the fragment...

```json
{
    "guid": "{65f3c7d5-f241-483c-b7c4-982b5361c318}",
    "name": "Ubuntu 28.02 Xylophonetic Xorn",
}
```

...and...

```json
{
    "updates": "{65f3c7d5-f241-483c-b7c4-982b5361c318}",
    "name": "Ubuntu 28.02 Xylophonetic Xorn",
}
```

The first one will create a new profile, even if the WSL distro generator didn't attest it.

We could use `source`, but that's baking a contract into another static file and holds Ubuntu back if we or a community member eventually adds "Microsoft.Terminal.WslButBetterBecauseItCanFindIconsButTheUserMayWantToDisableItAndFallBackToTheOldOneBecauseItIsExpensive" (sorry, mouthful).

We don't have a good story for sources overriding eachother, and we don't have a good system for overriding _regardless of source_.

This is an implementation of, effectively, "If-Match" from HTTP. "If you have one, I am interested. If you do not, leave me alone."

Reply:
AHHH okay. I see how that works. So for apps where we've already got a dynamic profile generator (WSL distros), they're going to want to use `updates`, but pretty much everyone else will want to use `guid`. Got it.